### DISCUSSION TITLE
Questions about implementing environment simulator of reinforcement learning task in LAVA

### DISCUSSION THREAD

**Discussion by CloudyDory** [lava-nc/lava#519] [Q&A]
Hi, I am working on a reinforcement learning task and I hope to implement the trained policy on Loihi. Like many reinforcement tasks, the environment receives the action from the agent and outputs an image as the next state. After reading the tutorials of LAVA, I guess that the environment simulator should also be wrapped as a LAVA process. And I have some questions about the implementation:

1. The environment simulator should run on CPU (ideally GPU), so I should use `PyAsyncProcessModel` and `AsyncProtocol`, is that correct?
2. It seems that there aren't many tutorials and documentations on `PyAsyncProcessModel` and `AsyncProtocol`. How to transfer data between `PyAsyncProcessModel` and `PyLoihiProcessModel`? Is it the same as how we setup ports and transfer data between two `PyLoihiProcessModel`?
3. It seems that the `run_async` method of `PyAsyncProcessModel` runs forever. How to make it pause when waiting for the next action input, and stop when the episode terminates?

**Comment by weidel-p** [comment]
Hi, thanks for your question.

As you know, you currently have two option on how to run your environmental process.
1. as PyAsyncProcessModel using the AsyncProtocol
2. as PyLoihiProcessModel using the LoihiProtocol

Although, I think it can make a lot of sense to run the environment asynchronously from the agent, your third point actually suggests that you want to synchronize the execution of the agent and the environment. So, if there is no specific reason to run the environment asynchronously, I would suggest to wrap it into a PyLoihiProcessModel and implement the dynamics in the run_spk function.

Otherwise, if you use the PyAsyncProcessModel, you can call the `pause` and `stop` function from your Process to pause and stop execution.

**Reply by CloudyDory** [reply]
Hi, thanks for the reply! 

If I continue to use the `PyAsyncProcessModel`, how to I specify `condition` and `run_cfg`? Currently I set `condition` to be `RunSteps(num_steps=10)` and `run_cfg` to be `LoihiSimCfg()`, but it seems that they are intended to be used in `PyLoihiProcessModel`.

And how should I use the pause() and stop() function within Process? Is there a code example for this?

**Comment by weidel-p** [comment]
You can use `RunContinuous` instead of `RunSteps` and then call `pause()` and `stop()` on the same Process you called `run()`.

You find examples here:
https://github.com/lava-nc/lava/blob/main/tutorials/in_depth/tutorial04_execution.ipynb
